### ISSUE TITLE
Entrypoint redirect to https rule breaks frontend rules

### ISSUE THREAD

**Issue by iadknet** [containous/traefik#679]
I created a docker-compose project to recreate this issue
https://github.com/iadknet/traefik-ssl-redirect

```
docker-compose up
```

Make an http request to the /test_app and get redirected to https (the redirect works fine)

```
$ curl -L -k -v http://localhost/test_app
*   Trying ::1...
* connect to ::1 port 80 failed: Connection refused
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /test_app HTTP/1.1
> Host: localhost
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 302 Found
< Server: nginx/1.11.4
< Date: Thu, 15 Sep 2016 23:08:10 GMT
< Content-Type: text/plain; charset=utf-8
< Content-Length: 5
< Connection: keep-alive
< Location: https://localhost/test_app
<
* Ignoring the response-body
* Connection #0 to host localhost left intact
* Issue another request to this URL: 'https://localhost/test_app'
* Found bundle for host localhost: 0x7f8d03e00790
*   Trying ::1...
* connect to ::1 port 443 failed: Connection refused
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 443 (#1)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate: dev.evisions
> GET /test_app HTTP/1.1
> Host: localhost
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: nginx/1.11.4
< Date: Thu, 15 Sep 2016 23:08:10 GMT
< Content-Type: text/plain; charset=utf-8
< Content-Length: 0
< Connection: keep-alive
<
* Connection #1 to host localhost left intact
```

The final request returns a 200, but the content is empty.

nginx logs:

```
nginx_1    | 172.19.0.1 - - [15/Sep/2016:23:08:10 +0000] "GET /test_app HTTP/1.1" 302 5 "-" "curl/7.43.0" "-"
nginx_1    | 172.19.0.1 - - [15/Sep/2016:23:08:10 +0000] "GET /test_app HTTP/1.1" 200 0 "-" "curl/7.43.0" "-"
```

traefik logs:

```
172.19.0.2 - - [15/Sep/2016:23:08:10 +0000] "GET /test_app HTTP/1.0" 0 0 "" "curl/7.43.0" 6 "" "" 90.389µs
```

To test without the redirect rules in place:

Open `traefik/traefik.toml` and comment out the following lines:

```
     [entryPoints.http.redirect]
       regex = "^http://(.*)"
       replacement = "https://$1"
```

```
docker-compose down && docker-compose up
```

Curl https://localhost/ and get the full page:

```
$ curl -L -k -v https://localhost/test_app
*   Trying ::1...
* connect to ::1 port 443 failed: Connection refused
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate: dev.evisions
> GET /test_app HTTP/1.1
> Host: localhost
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: nginx/1.11.4
< Date: Thu, 15 Sep 2016 23:10:56 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 118
< Connection: keep-alive
< Etag: W/"76-ktOkqas+2XplZPRxS5r6LA"
< Vary: Accept
< X-Powered-By: Express
<
<!DOCTYPE html>
<html>
<head>


  <title>Test App</title>

</head>
<body>
  <h1>Test Application</h1>
</body>
</html>
* Connection #0 to host localhost left intact
```

nginx logs:

```
nginx_1    | 172.19.0.1 - - [15/Sep/2016:23:12:10 +0000] "GET /test_app HTTP/1.1" 200 118 "-" "curl/7.43.0" "-"
```

traefik logs:

```
172.19.0.3 - - [15/Sep/2016:23:12:10 +0000] "GET / HTTP/1.0" 200 118 "" "curl/7.43.0" 2 "web-test-service" "http://172.19.0.5:8080" 1.487498ms
```

**Comment by ThinkBriK** [comment]
Got exactly the same behaviour when trying to workaround issue #563 with a regext on the entrypoint:
```
defaultEntryPoints = ["http"]
[entryPoints]
  [entryPoints.http]
  address = ":80"
    [entryPoints.http.redirect]
      regex = "(http:\\/\\/[^\\/]+\\/([^\\?]+)[^\\/])$"
      replacement = "$1/"
```
The regex is OK and redirection is OK too and I got the same empty 200 page.

I thing this regex could provide a workaround for #563 if only this issue is solved !

**Comment by avarabyeu** [comment]
+1, have the same problem. Trying to redirect from root to some endpoint:
```
[entryPoints.http.redirect]
        regex = "(http:\\/\\/[^\\/]+)\\/?$"
        replacement = "$1/ui"
```

**Comment by bernardoVale** [comment]
Guys, I need this fix for production. Any ideas where should we look in the code?

**Comment by bernardoVale** [comment]
Hello folks, I could reproduce the problem inside traefik dependency that is responsible for this problem. Take a look [here](https://github.com/vulcand/vulcand/issues/320)

**Comment by codymwalker** [comment]
+1, experiencing the same problem here.

**Comment by argeiger** [comment]
+1, hit the same issue

**Comment by chaosaffe** [comment]
+1, we have the same problem. Can someone point us to the relevant code so we can diagnose?

**Comment by bernardoVale** [comment]
@chaosaffe , a few parts of the code where I already debugged:

[Where the middlewrite is created](https://github.com/containous/traefik/blob/master/server.go#L799)

[Rewrite middleware implementation](https://github.com/containous/traefik/blob/master/middlewares/rewrite.go)

[Vendor HTTPServe implementation](https://github.com/vulcand/vulcand/blob/master/plugin/rewrite/rewrite.go#L65)

Hope this info help, you'll be my hero if you solve this problem. I really this rewrite middleware.

**Comment by argeiger** [comment]
Took a deeper dive into the code and I think the problem is in the following area:

`if entryPoint.Redirect != nil {
...
} else {
...
}`

If you have a redirect defined, you will not have a route defined to map the frontend with the backend server.  Looking through the debug log here is output when no regex is defined:

`DEBU[0000] Creating frontend frontend-app-kibana        
DEBU[0000] Wiring frontend frontend-app-kibana to entryPoint http 
DEBU[0000] Creating route route-host-app-kibana Host:app-kibana.service.cloudeng.stratus.apim.ibmcloud.com 
DEBU[0000] Creating entryPoint redirect http ->  : (.+)3eb01f4a-4267-4dad-b72a-eeaf1bd13b73(.+) -> http://yahoo.com 
DEBU[0000] Creating frontend frontend-gws-director      
DEBU[0000] Wiring frontend frontend-gws-director to entryPoint http 
DEBU[0000] Creating route route-host-gws-director PathPrefixStrip:/gws/director 
INFO[0000] Server configuration reloaded on :8777`

Here is when there is no regex defined:
`DEBU[0000] Creating frontend frontend-gws-director      
DEBU[0000] Wiring frontend frontend-gws-director to entryPoint http 
DEBU[0000] Creating route route-host-gws-director PathPrefixStrip:/gws/director 
DEBU[0000] Creating backend backend-gws-director        
DEBU[0000] Creating load-balancer wrr                   
DEBU[0000] Creating server server-gws_director-1e3c6656-f232-11e6-a19f-06d5bc01558b at http://172.17.0.3:31771 with weight 0 
DEBU[0000] Creating server server-gws_director-1e3c8d67-f232-11e6-a19f-06d5bc01558b at http://172.17.0.2:31573 with weight 0 
DEBU[0000] Creating server server-gws_director-1e3cdb88-f232-11e6-a19f-06d5bc01558b at http://172.17.0.7:31816 with weight 0 `

Note that one includes the Creating server which maps it to the backend.  I tried taking the else statement to have that code execute everytime and it mapped to the backend but didn't do the redirect regex/replacment.  Trying to dive deeper but have little time but this is what I have seen so far.

**Comment by argeiger** [comment]
Think I have a fix for it.  Will open up a PR.  If you have time, please help verify

**Comment by emilevauge** [comment]
Fixed by #1417
